anvil/src/state.rs

```diff
@@ -616,40 +620,7 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
         let keyboard_shortcuts_inhibit_state = KeyboardShortcutsInhibitState::new::<Self>(&dh);
 
         #[cfg(feature = "xwayland")]
-        let xwayland = {
-            XWaylandKeyboardGrabState::new::<Self>(&dh);
-
-            let (xwayland, channel) = XWayland::new(&dh);
-            let dh = dh.clone();
-            let ret = handle.insert_source(channel, move |event, _, data| match event {
-                XWaylandEvent::Ready {
-                    connection,
-                    client,
-                    client_fd: _,
-                    display,
-                } => {
-                    let mut wm = X11Wm::start_wm(data.handle.clone(), dh.clone(), connection, client)
-                        .expect("Failed to attach X11 Window Manager");
-                    let cursor = Cursor::load();
-                    let image = cursor.get_image(1, Duration::ZERO);
-                    wm.set_cursor(
-                        &image.pixels_rgba,
-                        Size::from((image.width as u16, image.height as u16)),
-                        Point::from((image.xhot as u16, image.yhot as u16)),
-                    )
-                    .expect("Failed to set xwayland default cursor");
-                    data.xwm = Some(wm);
-                    data.xdisplay = Some(display);
-                }
-                XWaylandEvent::Exited => {
-                    let _ = data.xwm.take();
-                }
-            });
-            if let Err(e) = ret {
-                tracing::error!("Failed to insert the XWaylandSource into the event loop: {}", e);
-            }
-            xwayland
-        };
+        let xwayland_shell_state = xwayland_shell::XWaylandShellState::new::<Self>(&dh.clone());
```

Comment:
I guess we could move this into start_xwayland, just like the XwaylandKeyboardGrab protocol?

Reply:
It means making that field an `Option`. What if I just move `XwaylandKeyboardGrab` protocol out here?

Reply:
sure

Reply:
Done.